libs/utils/src/properties_encoding.c

```diff
@@ -0,0 +1,580 @@
+            if (rc != 0) {
+                celix_err_push("Failed to set json object");
+                return ENOMEM;
+            }
+        }
+
+        jsonObj = subObj;
+        fieldName = slash + 1;
+        slash = strstr(fieldName, "/");
+    }
+
+    json_t* value;
+    celix_status_t status = celix_properties_entryValueToJson(fieldName, entry, flags, &value);
+    if (status != CELIX_SUCCESS) {
+        return status;
+    }
+    return celix_properties_addJsonValueToJson(value, fieldName, jsonObj, flags);
+}
+
+celix_status_t celix_properties_saveToStream(const celix_properties_t* properties, FILE* stream, int encodeFlags) {
+    json_auto_t* root = json_object();
+    if (!root) {
+        celix_err_push("Failed to create json object");
+        return ENOMEM;
+    }
+
+    if (!(encodeFlags & CELIX_PROPERTIES_ENCODE_FLAT_STYLE) && !(encodeFlags & CELIX_PROPERTIES_ENCODE_NESTED_STYLE)) {
+        //no encoding flags set, default to flat
+        encodeFlags |= CELIX_PROPERTIES_ENCODE_FLAT_STYLE;
+    }
+
+    CELIX_PROPERTIES_ITERATE(properties, iter) {
+        celix_status_t status;
+        if (encodeFlags & CELIX_PROPERTIES_ENCODE_FLAT_STYLE) {
+            status = celix_properties_addPropertiesEntryFlatToJson(&iter.entry, iter.key, root, encodeFlags);
+        } else {
+            assert(encodeFlags & CELIX_PROPERTIES_ENCODE_NESTED_STYLE);
+            status = celix_properties_addPropertiesEntryToJson(&iter.entry, iter.key, root, encodeFlags);
+        }
+        if (status != CELIX_SUCCESS) {
+            return status;
+        }
+    }
+
+    size_t jsonFlags = JSON_COMPACT;
+    if (encodeFlags & CELIX_PROPERTIES_ENCODE_PRETTY) {
+        jsonFlags = JSON_INDENT(2);
+    }
+
+    int rc = json_dumpf(root, stream, jsonFlags);
+    if (rc != 0) {
+        celix_err_push("Failed to dump json object to stream.");
+        return CELIX_FILE_IO_EXCEPTION;
+    }
+    return CELIX_SUCCESS;
+}
+
+celix_status_t celix_properties_save(const celix_properties_t* properties, const char* filename, int encodeFlags) {
+    FILE* stream = fopen(filename, "w");
+    if (!stream) {
+        celix_err_pushf("Failed to open file %s.", filename);
+        return CELIX_FILE_IO_EXCEPTION;
+    }
+    celix_status_t status = celix_properties_saveToStream(properties, stream, encodeFlags);
+    fclose(stream);
+    return status;
+}
+
+celix_status_t celix_properties_saveToString(const celix_properties_t* properties, int encodeFlags, char** out) {
+    *out = NULL;
+    celix_autofree char* buffer = NULL;
+    size_t size = 0;
+    FILE* stream = open_memstream(&buffer, &size);
+    if (!stream) {
+        celix_err_push("Failed to open memstream.");
+        return ENOMEM;
+    }
+
+    celix_status_t status = celix_properties_saveToStream(properties, stream, encodeFlags);
+    fclose(stream);
+    if (status == CELIX_SUCCESS) {
+        *out = celix_steal_ptr(buffer);
+    }
+    return status;
```

Comment:
Shall we check `buffer == NULL`? From what I can tell from glibc's implementation:

```C
static void
_IO_mem_finish (FILE *fp, int dummy)
{
  struct _IO_FILE_memstream *mp = (struct _IO_FILE_memstream *) fp;

  *mp->bufloc = (char *) realloc (fp->_IO_write_base,
				  fp->_IO_write_ptr - fp->_IO_write_base + 1);
  if (*mp->bufloc != NULL)
    {
      (*mp->bufloc)[fp->_IO_write_ptr - fp->_IO_write_base] = '\0';
      *mp->sizeloc = fp->_IO_write_ptr - fp->_IO_write_base;

      fp->_IO_buf_base = NULL;
    }

  _IO_str_finish (fp, 0);
}
```

If `realloc` fails, `buffer` (*mp->bufloc) will be `NULL`.